bokeh/models/annotations.py

```diff
@@ -183,6 +186,100 @@ class BoxAnnotation(Annotation):
+
+    x = NumberSpec(help="""
+    The x-coordinates to locate the text anchors.
+    """)
+
+    x_units = Enum(SpatialUnits, default='data', help="""
+    The unit type for the xs attribute. Interpreted as "data space" units
+    by default.
+    """)
+
+    y = NumberSpec(help="""
+    The y-coordinates to locate the text anchors.
+    """)
+
+    y_units = Enum(SpatialUnits, default='data', help="""
+    The unit type for the ys attribute. Interpreted as "data space" units
+    by default.
+    """)
+
+    text = StringSpec("text", help="""
+    The text values to render.
+    """)
+
+    angle = AngleSpec(default=0, help="""
+    The angles to rotate the text, as measured from the horizontal.
+    """)
+
+    x_offset = NumberSpec(default=0, help="""
+    Offset values to apply to the x-coordinates.
+
+    This is useful, for instance, if it is desired to "float" text a fixed
+    distance in screen units from a given data position.
+    """)
+
+    y_offset = NumberSpec(default=0, help="""
+    Offset values to apply to the y-coordinates.
+
+    This is useful, for instance, if it is desired to "float" text a fixed
+    distance in screen units from a given data position.
+    """)
+
+    text_props = Include(TextProps, use_prefix=False, help="""
+    The %s values for the text.
+    """)
+
+    background_props = Include(FillProps, use_prefix=True, help="""
+    The %s values for the text bounding box.
+    """)
+
+    background_fill_color = Override(default="#ffffff")
+
+    background_fill_alpha = Override(default=0.0)
+
+    border_props = Include(LineProps, use_prefix=True, help="""
+    The %s values for the text bounding box.
+    """)
+
+    border_line_color = Override(default="black")
+
+    border_line_alpha = Override(default=0.0)
+
+    source = Instance(DataSource, help="""
+    Local data source to use when rendering annotations on the plot.
+    """)
+
+    x_range_name = String('default', help="""
+    A particular (named) x-range to use for computing screen locations when
+    rendering annotations on the plot. If unset, use the default x-range.
+    """)
+
+    y_range_name = String('default', help="""
+    A particular (named) y-range to use for computing screen locations when
+    rendering annotations on the plot. If unset, use the default y-range.
+    """)
+
+    render_mode = Enum(RenderMode, default="canvas", help="""
+    Specifies whether the text is rendered as a canvas element or as an
+    css element overlaid on the canvas. The default mode is "canvas".
+
+    .. warning::
+        Not all visual styling properties are supported if the render_mode is
+        set to "css". The border_line_dash property isn't fully supported and
+        border_line_dash_offset isn't supported at all. Setting text_alpha will
+        modify the opacity of the entire background box and border in addition
+        to the text. Finally, clipping Label annotations inside of the plot
+        area isn't supported in "css" mode.
```

Comment:
It may also be worth noting that css labels will not be present in the output from the "save" tool. I wouldn't put this is in the warning as I'd argue it's a feature not a bug.